lattice_controller_engine/lib/lattice_controller_engine/observed/lattice.ex

```diff
@@ -0,0 +1,474 @@
+  lattice events
+  """
+  alias __MODULE__
+  alias LatticeControllerEngine.Observed.{Provider, Host, Instance, LinkDefinition}
+
+  # We need the keys to be there, even if they hold empty lists
+  @enforce_keys [:actors, :providers, :hosts, :linkdefs]
+  defstruct [:actors, :providers, :hosts, :linkdefs, :ocimap, :instance_tracking]
+
+  @typedoc """
+  A provider key is the provider's public key accompanied by the link name
+  """
+  @type provider_key :: {String.t(), String.t()}
+
+  @type actormap :: %{required(String.t()) => [Instance.t()]}
+  @type providermap :: %{required(provider_key()) => Provider.t()}
+  @type hostmap :: %{required(String.t()) => Host.t()}
+  # map between OCI image URL/imageref and public key
+  @type ocimap :: %{required(String.t()) => String.t()}
+
+  @typedoc """
+  Keys are the instance ID, values are ISO 8601 timestamps in UTC
+  """
+  @type instance_trackmap :: %{required(String.t()) => DateTime.t()}
+
+  @typedoc """
+  The root structure of an observed lattice. An observed lattice keeps track
+  of the actors, providers, link definitions, and hosts within it.
+  """
+  @type t :: %Lattice{
+          actors: actormap(),
+          providers: providermap(),
+          hosts: hostmap(),
+          linkdefs: [LinkDefinition.t()],
+          instance_tracking: instance_trackmap(),
+          ocimap: ocimap()
+        }
+
+  @spec new :: t()
+  def new() do
+    %Lattice{
+      actors: %{},
+      providers: %{},
+      hosts: %{},
+      linkdefs: [],
+      instance_tracking: %{},
+      ocimap: %{}
+    }
+  end
+
+  @spec apply_event(
+          t(),
+          Cloudevents.Format.V_1_0.Event.t()
+        ) :: t()
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{"actors" => _actors, "providers" => _providers, "labels" => labels},
+          datacontenttype: "application/json",
+          source: source_host,
+          time: stamp,
+          type: "com.wasmcloud.lattice.host_heartbeat"
+        }
+      ) do
+    record_host(l, source_host, labels, stamp)
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "public_key" => public_key,
+            "link_name" => link_name,
+            "annotations" => annotations
+          },
+          datacontenttype: "application/json",
+          source: source_host,
+          type: "com.wasmcloud.lattice.health_check_passed"
+        }
+      ) do
+    l
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "public_key" => public_key,
+            "link_name" => link_name,
+            "annotations" => annotations
+          },
+          datacontenttype: "application/json",
+          source: source_host,
+          type: "com.wasmcloud.lattice.health_check_failed"
+        }
+      ) do
+    l
+  end
```

Comment:
This feels like something we should keep track of (health check status) but I'm fine if this is still a TODO. Just making sure we aren't doing nothing here on purpose

Reply:
yes we'll ultimately want to do something meaningful with the health check pass fail. I've added this in #3